### PR TITLE
Add mouse driver/ps2 functions

### PR DESCRIPTION
**Pull request by alexmwu**
Add general ps2 port handling functions.
Add mouse driver.
Add mouse drawing functionality with full erasure.

### CODE REVIEW

File: src/graphics.h [hunk 1]
```diff
@@ -83,6 +83,13 @@ void graphics_arc(int x, int y, double r, double start_theta, double end_theta,
  * @param graphics_color The color to draw in
  */
 void graphics_circle(int x, int y, double r, struct graphics_color c);
+
+/**
+ * @brief draw the mouse
```
Comment (kkoser) [inline src/graphics.h]: np: capital D

File: src/keyboard.c [hunk 2]
```diff
@@ -47,7 +47,7 @@ static int keyboard_scan() {
     int code;
     int ack;
 
-    code = inb(KEYBOARD_PORT_A);
+    code = inb(PS2_DATA_PORT);
     iowait();
     ack = inb(KEYBOARD_PORT_B);
```
Comment (kkoser) [inline src/keyboard.c:52]: why didn't we change KEYBOARD_PORT_B to a PS2 port?

File: src/main.c [hunk 3]
```diff
@@ -9,7 +9,9 @@ See the file LICENSE for details.
 #include "memory_raw.h"     // memory_init
 #include "process.h"
 #include "interrupt.h"
+#include "ps2.h"
```
Comment (kkoser) [inline src/main.c]: unnecessary include

File: src/memory_raw.c [hunk 4]
```diff
@@ -195,3 +195,20 @@ static int memory_is_range_type_available(int type) {
         return 0;
     }
 }
+
+int memory_freemap_walk() {
```
Comment (kkoser) [inline src/memory_raw.c]: ...Why is this here?

Comment (alexmwu) [inline src/memory_raw.c]: forgot about this. I used it for figuring out which line caused the "memory all allocated error"

File: src/mouse.c [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+#include "process.h"
+#include "graphics.h"
+
+static uint8_t mouse_cycle = 0;
+static uint8_t mouse_byte[3];
+struct graphics_color mouse_fg_color = {0, 255, 0};
+
+static struct list queue = { 0, 0 };
+
+int mouse_scan() {
```
Comment (kkoser) [inline src/mouse.c:34]: Can we get some comments for the internal functions? This goes for all of them

File: src/keyboard.c (hunk 2 above)
Comment (alexmwu) [inline src/keyboard.c:52]: Thain used it. not a specifically PS2 port. From OS Dev Wiki: 

Historical note : The PC-XT PPI had used port 0x61 to reset the keyboard interrupt request signal (among other unrelated functions). Port 0x61 has no keyboard related functions on AT and PS/2 compatibles.

File: src/mouse.c [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+			break;
+		case 1:
+			mouse_byte[1] = inb(PS2_DATA_PORT);
+			mouse_cycle++;
+			break;
+		case 2:
+			mouse_byte[2] = inb(PS2_DATA_PORT);
+			mouse_cycle++;
+			mouse_cycle = 0;
+            mouse_map();
```
Comment (kkoser) [inline src/mouse.c]: Incorrect indentation

**Comment by kkoser** [comment]
Please use spaces, not tabs

File: src/mouse.c [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            graphics_mouse();
+			break;
+	}
+
+    process_wakeup(&queue);
```
Comment (kkoser) [inline src/mouse.c]: Is this necessary / effective? I'm also confused how this works when queue is always empty

File: src/mouse.c [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    ps2_command_write(0xD4, 0xF6);
+    ps2_controller_read_ready();
+    int ack = inb(PS2_DATA_PORT);
+	if (ack != MOUSE_ACK) {
+		// TODO: handle bad ack
+		console_printf("mouse: bad ack: %x\n", ack);
+	}
+	// enable packet streaming
+	ps2_command_write(0xD4, 0xF4);
+    ps2_controller_read_ready();
```
Comment (kkoser) [inline src/mouse.c:153]: Inconsistent indentation

File: src/mouse.c [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+	// enable port 2
+    cont_config_byte |= 0x02;
+    // clear bit 5 (disable mouse clock)
+    cont_config_byte &= ~0x08;
+    ps2_write_controller_config_byte(cont_config_byte);
+	// send set defaults command to mouse
+    ps2_command_write(0xD4, 0xF6);
+    ps2_controller_read_ready();
+    int ack = inb(PS2_DATA_PORT);
+	if (ack != MOUSE_ACK) {
```
Comment (kkoser) [inline src/mouse.c]: inconsistent indentation

File: src/mouse.c [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+		// TODO: handle bad ack
+		console_printf("mouse: bad ack: %x\n", ack);
+	}
+	// enable packet streaming
+	ps2_command_write(0xD4, 0xF4);
+    ps2_controller_read_ready();
+    ack = inb(PS2_DATA_PORT);
+	if (ack != MOUSE_ACK) {
```
Comment (kkoser) [inline src/mouse.c]: indentation

File: src/ps2.c [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        second_channel_enabled = 1;
+    }
+
+    ps2_command_write(0x60, cont_config_byte);
+    ps2_write_controller_config_byte(cont_config_byte);
+
+    console_printf("free?: %d\n", memory_freemap_walk());
+    // causes memory issues (the outb call)
+    /*
+    // test PS/2 controller (causes memory issues)
```
Comment (kkoser) [inline src/ps2.c]: We should remove all of this commented out code

File: src/ps2.c [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    cont_config_byte &= ~(0x01 | 0x02 | 0x40);
+    // if clear, not dual channel PS/2 controller (b/c second PS/2 port disabled)
+    if (((cont_config_byte >> 5) & 0x01) == 1) {
+        second_channel_enabled = 1;
+    }
+
+    ps2_command_write(0x60, cont_config_byte);
+    ps2_write_controller_config_byte(cont_config_byte);
+
+    console_printf("free?: %d\n", memory_freemap_walk());
```
Comment (kkoser) [inline src/ps2.c]: Is this still necessary? Looks like a debugging tool

File: src/ps2.h [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+#define PS2_H
+
+#define PS2_DATA_PORT 0x60
+#define PS2_STATUS_REGISTER 0x64
+#define PS2_COMMAND_REGISTER 0x64
+#define PS2_TIMEOUT 1
+
+#include "ioports.h"
+
+void ps2_init();
```
Comment (kkoser) [inline src/ps2.h]: This function is never used, do we need it?

File: src/ps2.h [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * @brief whether second channel is enabled
+ * @details This function returns (default no) whether the second PS2 channel is enabled.
+ * @return boolean on if second channel enabled
+ */
+int ps2_second_channel_enabled();
```
Comment (kkoser) [inline src/ps2.h:22]: This function is never used, do we need it?

**Comment by kkoser** [comment]
Tested and confirmed, mouse draws on screen

File: src/ps2.h (hunk 13 above)
Comment (alexmwu) [inline src/ps2.h]: probably not. if mouse_init works without using ps2_init, we can probably ignore the tests and initialization code in ps2_init

File: src/ps2.h (hunk 14 above)
Comment (alexmwu) [inline src/ps2.h:22]: if no ps2_init, this is useless.

**Comment by jhamilt5** [comment]
Mouse drew on my screen too. You need to fix your merge conflict in Makefile, though.

**Comment by alexmwu** [comment]
fixed merge conflicts

File: src/mouse.c [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+        case 2:
+            mouse_byte[2] = inb(PS2_DATA_PORT);
+            mouse_cycle++;
+            mouse_cycle = 0;
+            mouse_map();
+            graphics_mouse();
+            break;
+    }
+
+    process_wakeup(&queue);
```
Comment (kkoser) [inline src/mouse.c]: Im still confused by this. You are providing an empty queue, to which you never add anything. What does this do?

File: src/mouse.h [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+#ifndef MOUSE_H
+#define MOUSE_H
+
+#include "graphics.h"
+
+#define MOUSE_ACK 0xFA
+// TODO: to support hotplugging, then every so often, need to query if mouse is alive (send 0xEB and recv and ack)
+extern struct graphics_color mouse_fg_color;
+
+int mouse_x;
```
Comment (kkoser) [inline src/mouse.h:18]: This allows people to set the mouse x/y themselves. Is this something we want, or do we want to hide these values behinds getter functions? I could see a use case for it but it just seems a little dangerous

**Comment by alexmwu** [comment]
any idea why this says the branch is out-of-date? If I click update branch, it just gives a merge commit with no changes

**Comment by kkoser** [comment]
I would imagine its because this branch is out of date, but has no merge conflicts, and so it can be automatically fast forwarded

**Comment by alexmwu** [comment]
so should I just update branch and have a useless merge commit?

**Comment by rboccabe** [comment]
Sure. Whatever it takes to get this in. Useless commit is fine except @shuyangli will roll over in his grave.

**Comment by kkoser** [comment]
Tested and confirmed, we have a working mouse!

File: src/graphics.c [hunk 17]
```diff
@@ -172,3 +188,92 @@ void graphics_char(int x, int y, char ch, struct graphics_color fgcolor,
     return graphics_bitmap(x, y, FONT_WIDTH, FONT_HEIGHT, &fontdata[u],
                            fgcolor, bgcolor);
 }
+
+uint8_t graphics_verify_color_range(uint8_t x) {
```
Comment (kkoser) [inline src/graphics.c]: This function is redundant, an unsigned 8 bit integer can only hold values between 0 and 255

File: src/graphics.h [hunk 18]
```diff
@@ -91,4 +92,28 @@ void graphics_bitmap(int x, int y, int width, int height, uint8_t * data,
...
                      struct graphics_color bgcolor);
 
+/**
+ * @brief Ensures that a color is within 0 and 255.
+ * @param x A color value.
+ * @return A clipped color value, between 0 and 255.
+ */
+uint8_t graphics_verify_color_range(uint8_t x);
+
+void graphics_copy_to_color_buffer(int x, int y, int width, int height, struct graphics_color *buffer, int buf_size);
```
Comment (kkoser) [inline src/graphics.h:105]: Doxygen comments plz. Also for counterpart function

File: src/ps2.h (hunk 13 above)
Comment (kkoser) [inline src/ps2.h]: So are we removing this?

File: src/graphics.c (hunk 17 above)
Comment (alexmwu) [inline src/graphics.c]: Yes, but ints below 0 and above 255 would get casted and we want behavior where they get bounded, not casted. This actually used to be in console. Seemed more appropriate in graphics.

Comment (kkoser) [inline src/graphics.c]: The parameter for this function (and the function it is used in) is a uin8_t, meaning that the cast has already happened. I think we are trying to do too much here -- the parameter type for color functions is already unit8_t, indicating that the values are between 0 and 255. This seems like enough to me.

File: src/console.c [hunk 19]
```diff
@@ -97,30 +97,21 @@ void console_init() {
...
-    return x;
-}
-
 void console_set_fgcolor(uint8_t r, uint8_t g, uint8_t b) {
-    r = console_verify_color_range(r);
-    g = console_verify_color_range(g);
-    b = console_verify_color_range(b);
+    r = graphics_verify_color_range(r);
+    g = graphics_verify_color_range(g);
+    b = graphics_verify_color_range(b);
```
Comment (rboccabe) [inline src/console.c]: Doubles up with @kkoser 's earlier comment regarding the verification. I like the idea of the console asking graphics, but right now nothing different will ever be coming back if console_set_fg_color has uint8_t's coming in. It's 3 extra function calls that would allow us to change graphics_verify later with no impact here, but I don't see us changing our color ranges.

File: src/graphics.c [hunk 20]
```diff
@@ -47,6 +48,21 @@ static inline void plot_pixel(int x, int y, struct graphics_color c) {
...
+    // if the point is in the mouse region, draw to the mouse buffer, copy into the video region, and draw the mouse
+    if (mouse_inited && diff_x < MOUSE_SIDE_2 && diff_x > -MOUSE_SIDE_2 && diff_y < MOUSE_SIDE_2 && diff_y > -MOUSE_SIDE_2) {
+        // wraps to [0, 28]
+        int x_index = diff_x + MOUSE_SIDE_2 - 1;
+        int y_index = diff_y + MOUSE_SIDE_2 - 1;
+        int index = y_index * (MOUSE_SIDE - 1) + x_index;
+        mouse_draw_buffer[index].r = c.r;
+        mouse_draw_buffer[index].g = c.g;
+        mouse_draw_buffer[index].b = c.b;
+        graphics_draw_mouse();
```
Comment (rboccabe) [inline src/graphics.c:71]: If we are just managing the buffer behind the mouse here, do we need to draw the mouse again?

File: src/memory_raw.c [hunk 21]
```diff
@@ -195,3 +195,4 @@ static int memory_is_range_type_available(int type) {
         return 0;
     }
 }
+
```
Comment (rboccabe) [inline src/memory_raw.c]: np: please remove so we don't unnecessarily change a file.

File: src/graphics.h [hunk 22]
```diff
@@ -83,6 +83,7 @@ void graphics_arc(int x, int y, double r, double start_theta, double end_theta,
  * @param graphics_color The color to draw in
  */
 void graphics_circle(int x, int y, double r, struct graphics_color c);
+
```
Comment (rboccabe) [inline src/graphics.h]: np: any reason for an extra line here?

File: src/graphics.h [hunk 23]
```diff
@@ -91,4 +92,28 @@ void graphics_bitmap(int x, int y, int width, int height, uint8_t * data,
                      struct graphics_color fgcolor,
                      struct graphics_color bgcolor);
 
+/**
+ * @brief Ensures that a color is within 0 and 255.
+ * @param x A color value.
```
Comment (rboccabe) [inline src/graphics.h]: np: extra line here as per style guide

File: src/graphics.c (hunk 20 above)
Comment (alexmwu) [inline src/graphics.c:71]: if we draw under the mouse, we need to draw into the buffer (and onto video) and then draw the mouse over. Although, I did just realize we need to move the draw to video before draw to buffer and draw mouse.

Comment (rboccabe) [inline src/graphics.c:71]: Why not just put the draw to the video buffer in an else block? If the pixel we are drawing is under the mouse, just update the buffer behind the mouse. Else, actually update the video buffer. I don't see why we would always update the video buffer, then sometimes fill the mouse buffer and redraw the mouse over on the video buffer.

Comment (alexmwu) [inline src/graphics.c:71]: Because if we just update the mouse buffer, the video buffer does not get updated.

**Comment by alexmwu** [comment]
Fixed comments. let me know of any other issues.

File: src/graphics.c (hunk 20 above)
Comment (alexmwu) [inline src/graphics.c:71]: Unless you have strong objections to what I do here (basically look at a square region rather than just the mouse imprint), I think getting this in is more important. I can add an issue in Jira to optimize it. @rboccabe

File: src/graphics.c [hunk 24]
```diff
@@ -172,3 +188,82 @@ void graphics_char(int x, int y, char ch, struct graphics_color fgcolor,
...
+    for (j = y; j < y + height; j++) {
+        for (i = x; i < x + width; i++) {
+            if (i < 0 || i > video_xres - 1 || j < 0 || j > video_yres - 1) {
+                continue;
+            }
+            uint8_t *v = video_buffer + video_xbytes * j + i * 3;
+            v[2] = buffer[buf_ix].r;
+            v[1] = buffer[buf_ix].g;
+            v[0] = buffer[buf_ix].b;
+			buf_ix++;
```
Comment (kkoser) [inline src/graphics.c]: np: weird indentation

File: src/graphics.c [hunk 25]
```diff
@@ -172,3 +188,82 @@ void graphics_char(int x, int y, char ch, struct graphics_color fgcolor,
...
+    for (j = y; j < y + height; j++) {
+        for (i = x; i < x + width; i++) {
+            if (i < 0 || i > video_xres - 1 || j < 0 || j > video_yres - 1) {
+                continue;
+            }
+            uint8_t *v = video_buffer + video_xbytes * j + i * 3;
+            buffer[buf_ix].r = v[2];
+            buffer[buf_ix].g = v[1];
+            buffer[buf_ix].b = v[0];
+			buf_ix++;
```
Comment (kkoser) [inline src/graphics.c]: np: weird indentation

File: src/graphics.h [hunk 26]
```diff
@@ -91,4 +91,43 @@ void graphics_bitmap(int x, int y, int width, int height, uint8_t * data,
...
+ * @param buffer Pointer to a struct graphics_color buffer.
+ * @param buf_size Size of the buffer.
+ */
+void graphics_copy_from_color_buffer(int x, int y, int width, int height, struct graphics_color *buffer, int buf_size);
+
+/**
+ * @brief Draws mouse on the vid buf
+ * @details This only draws the mouse on the vid buf. It does not look at the buffer underneath it.
+ */
+void graphics_draw_mouse();
```
Comment (kkoser) [inline src/graphics.h]: Does this function need to be public, or could we remove the declaration here?

File: src/graphics.c [hunk 27]
```diff
@@ -61,12 +66,7 @@ static inline void plot_pixel(int x, int y, struct graphics_color c) {
         mouse_draw_buffer[index].r = c.r;
         mouse_draw_buffer[index].g = c.g;
         mouse_draw_buffer[index].b = c.b;
-        graphics_draw_mouse();
```
Comment (rboccabe) [inline src/graphics.c]: Did you ever replace this graphics_draw_mouse( ), or should it entirely disappear?

Comment (alexmwu) [inline src/graphics.c]: No, that was an accident.

File: src/graphics.h (hunk 26 above)
Comment (alexmwu) [inline src/graphics.h]: I thought it may be useful to make the raw draw public, but I can't think of any uses. I can remove it

**Comment by jhamilt5** [comment]
It's rebased, huzzah! LGTM
